pageserver/src/relish_storage/rust_s3.rs

```diff
@@ -0,0 +1,195 @@
+                SEPARATOR = S3_FILE_SEPARATOR
+            ),
+            None => S3_FILE_SEPARATOR.to_string(),
+        }
+    }
+
+    fn object_path_from_key(&self, key: String) -> Result<RelishStoragePath, String> {
+        if let Some(tenant_id) = &self.tenant_id {
+            if !key.starts_with(&tenant_id.to_string()) {
+                return Err(format!("Received an object with key {} that does not start with the preset tenant id prefix {}", key, tenant_id));
+            }
+        }
+
+        let (name, extension) = name_and_extension(&key)
+            .ok_or_else(|| format!("Failed to deternime name and extension for key {}", key))?;
+        let key_length = key.len();
+
+        Ok(RelishStoragePath {
+            file_name_offset: key_length
+                - name.len()
+                - extension.map(|extension| extension.len() + 1).unwrap_or(0),
+            extension_offset: extension.map(|extension| key_length - extension.len()),
+            key,
+        })
+    }
+}
+
+impl RelishStorage for RustS3 {
+    type Error = S3Error;
+
+    fn list_relishes(&self) -> Result<Vec<RelishStoragePath>, Self::Error> {
+        let list_response = UPLOAD_RUNTIME
+            .block_on(async { self.bucket.list(self.root_path(), None).await })
+            .map_err(|e| S3Error(format!("Failed to list s3 objects: {}", e)))?;
+
+        Ok(list_response
+            .into_iter()
+            .flat_map(|response| response.contents)
+            .filter_map(|s3_object| self.object_path_from_key(s3_object.key).ok())
+            .collect())
+    }
+
+    fn download_relish(&self, path: &RelishStoragePath) -> Result<Vec<u8>, Self::Error> {
+        let (data, code) = UPLOAD_RUNTIME
+            .block_on(async { self.bucket.get_object(&path.key).await })
+            .map_err(|e| {
+                S3Error(format!(
+                    "Failed to download s3 object with key '{}': {}",
+                    path.key, e
+                ))
+            })?;
+        if code != 200 {
+            Err(S3Error(format!(
+                "Received non-200 exit code during downloading object from directory, code: {}",
+                code
+            )))
+        } else {
+            Ok(data)
+        }
+    }
+
+    fn delete_relish(&self, path: &RelishStoragePath) -> Result<(), Self::Error> {
+        let (_, code) = UPLOAD_RUNTIME
+            .block_on(async { self.bucket.delete_object(&path.key).await })
+            .map_err(|e| {
+                S3Error(format!(
+                    "Failed to delete s3 object with key '{}': {}",
+                    path.key, e
+                ))
+            })?;
+        if code != 200 {
+            Err(S3Error(format!(
+                "Received non-200 exit code during deleting object with key '{}', code: {}",
+                path.key, code
+            )))
+        } else {
+            Ok(())
+        }
+    }
+
+    // TODO kb store in the same dir structure as on fs?
```

Comment:
It feels like we'd better maintain the same directory structure, as we do now locally, should we do that?